### PR TITLE
607_merge_experiment: Implement merge functions

### PR DESCRIPTION
**Pull request by bundfussr**

### CODE REVIEW

File: tests/testthat/test-derive_merged.R
Comment (konstantinakoukR) [inline tests/testthat/test-derive_merged.R:159]: lint fails at this line, in case you missed that :)

Comment (bundfussr) [reply tests/testthat/test-derive_merged.R:159]: Thanks, I'll fix it together with the replacement of the internal `left_join()` calls.

File: R/derive_merged.R
Comment (kabis-ops) [inline R/derive_merged.R:225]: It sounds strange to me to use the derived variable as input to the function.
Should it refers only to variable in dataset_add and then inside the function, the derived variable is used in addition to the specified var?
(just thinking)
` order = vars(EXTRT, EXSEQ)` 
and when you call derive_meged_vars you add the derived variable to the order var?: 
`order = vars(EXTRT, EXSEQ, TRTSDT)`

File: R/derive_merged.R
Comment (kabis-ops) [inline R/derive_merged.R:436]: Should add example with `missing_value`

File: R/derive_merged.R
Comment (kabis-ops) [inline R/derive_merged.R:495]: Is that needed? The `filter_add` param should do the filtering?

File: R/derive_merged.R
Comment (kabis-ops) [inline R/derive_merged.R:568]: false_value is given twice

File: R/derive_merged.R
Comment (yangc71) [inline R/derive_merged.R:83]: Can we add some clarification this is essentially a left join (e.g. all records from the left `dataset` are kept and records from the right `dataset_add` are not included if there are no matching records in the left `dataset`)?

File: R/derive_merged.R
Comment (yangc71) [reply R/derive_merged.R:38]: Maybe we can clarify `mode` is required (permitted value `"first"` and `"last"`) if `order` variable is provided, else `mode` is defaulted to `NULL` if `order` is not provided?

File: R/derive_merged.R
Comment (yangc71) [inline R/derive_merged.R:286]: When `date_imputation` is NULL and a partial date exists, then the `derive_vars_dt` step converts the partial dates to missing. As a result, the date variable specified in `order` dictates which record is chosen. For example, if there are two records with `EXSTDTC` being`“2020”` and `“2020-12-14”`, then converted `TRTSDT` are `NA` and `2020-12-14`. Assume `mode = “first”`, if `order = vars(TRTSDT)` (converted date) then the record with complete date is chosen. But if `order = vars(EXSTDTC)` (unconverted character date), then the record with partial date is chosen. I’m not sure if this is intended or I'm making this too confusing. If it's intended, should we include some clarifications about which records are chosen when partial dates are not imputed?

File: R/derive_merged.R
Comment (yangc71) [inline R/derive_merged.R:671]: I understand we created multiple merging functions to avoid users writing expressions in input. But in general for these wrapper functions, I'm wondering is it useful to allow users to keep more source variables (without transformation)? So they don't have to run merging multiple times if they want to merge in multiple variables (for example, from my understanding if they want to merge a converted character variable, but in addition also keep other variables as-is from `dataset_add` they have to run `derive_var_merged_character` then `derive_vars_merged`).

**Review by yangc71 — COMMENTED** [review]
Hi Stefan, apologies for the delayed review, I was a bit busy with study work this month. I have added my comments, and I know I missed several meetings, so please ignore any comments that have already been discussed or are not applicable.

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:4]: Yes, updated

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:13]: Added

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:38]: Updated

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:21]: Updated

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:46]: Updated

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:88]: `1.` is just a placeholder to indicate a numbered list. In the rendered version it is replaced by the correct number.

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:121]: For the merge functions the selection is based on `dataset_add` only.

I agree that we should have a function for your use case. We have already an issue for it (https://github.com/pharmaverse/admiral/issues/578).

I would not integrate this functionality into the merge functions because the flow is different and it requires much more resources. The datasets need to be joined first and then restricted. Joining first can lead to huge datasets. Thus it should be done only if it is really required.

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:30]: Added

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:1]: Updated

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:26]: I have updated the description. Is it clear now?

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:48]: Added

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:91]: I am not sure. "created" sounds a bit confusing to me. What about "are renamed (if requested) and merged"?

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:85]: Updated

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:459]: Good idea. Done

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:518]: In principle yes, but I wanted to distinguish between no data and condition not fulfilled.

I have added an example (and the `missing_value` parameter).

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:591]: This is intended the first one is for condition is false and the second one for condition is `NA`.

However, I have replaced the second one by `missing_value`.

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:18]: Fixed

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:139]: Good catch! `new_vars` added.

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:83]: I have added more details. Is it clear now?

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:286]: Good point! I have added some details to the description of the `order` parameter. I would expect that in most cases missings are excluded by `filter_add`.

However, the current behaviour looks a bit inconsistent. For `mode = "first"` observations with `NA` are selected only if all observation are `NA` while for `mode = "last"` observations with `NA` are always selected. Maybe we should change this behaviour. What do you think?

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:671]: You are right, with our current approach the user may need to call `derive_vars_merged_*` functions more than once for merging multiple variables. This was not my favourite solution but the majority preferred this approach. We have discussed it multiple times. Thus I would not like to discuss it again. At least not now. I would wait for user feedback before rediscussing it.

**Comment by bundfussr** [comment]
Hi @kabis-ops , @yangc71 , thanks for your review and all your comments. I hope I have addressed all of them.

File: R/derive_merged.R
Comment (yangc71) [reply R/derive_merged.R:114]: Yes, thanks!

File: R/derive_merged.R
Comment (yangc71) [reply R/derive_merged.R:345]: Thanks for adding the details in the description. 

If I'm understanding correctly, this behavior seems to be carried from `filter_extreme`. I think in some cases the users may not want to choose records with `NA`. Maybe we can add a parameter for `filter_extreme` (e.g. `remove_na`) to allow users to choose whether `NA` should be considered?

File: R/derive_merged.R
Comment (yangc71) [reply R/derive_merged.R:807]: I see. Thanks for the information.

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:345]: Yes, you are correct. I have added it to the agenda of the next "Questions and Comments" meeting as it affects many functions.

Comment (bundfussr) [reply R/derive_merged.R:345]: Hi @yangc71 , we discussed the topic today and decided to keep the behavior as it is because it is in line with the `arrange()` function. We also discussed to provide a `remove_na` parameter but decided not to provide it because usually `filter_extreme` is not called directly but within a derivation and the derivations already provide a `filter` parameter to exclude NAs.

File: R/derive_merged.R
Comment (kabis-ops) [inline R/derive_merged.R:27]: there is a missing parethesis after AVAL)

File: R/derive_merged.R
Comment (kabis-ops) [reply R/derive_merged.R:619]: OK, it may be good to add some details for the `condition` parameter, then?

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:27]: Thanks, fixed

File: R/derive_merged.R
Comment (bundfussr) [reply R/derive_merged.R:619]: Yes, I have added a description. I have also updated the details section as the `missing_value` parameter was not mentioned there.

**Review by yangc71 — APPROVED** [review]

**Pull request merged**
